File: testdata/jennies/builders/struct_with_defaults/TypescriptBuilder/src/struct_with_defaults/nestedstruct_builder_gen.ts
import * as cog from '../cog';
import * as struct_with_defaults from '../struct_with_defaults';

export class NestedStructBuilder implements cog.Builder<struct_with_defaults.NestedStruct> {
    private readonly internal: struct_with_defaults.NestedStruct;

    constructor() {
        this.internal = struct_with_defaults.defaultNestedStruct();
    }

    build(): struct_with_defaults.NestedStruct {
        return this.internal;
    }

    stringVal(stringVal: string): this {
        this.internal.stringVal = stringVal;
        return this;
    }

    intVal(intVal: number): this {
        this.internal.intVal = intVal;
        return this;
    }
}

File: testdata/jennies/builders/struct_with_defaults/GoBuilder/structwithdefaults/nestedstruct_builder_gen.go
package structwithdefaults

import (
	cog "github.com/grafana/cog/generated/cog"
)

var _ cog.Builder[NestedStruct] = (*NestedStructBuilder)(nil)

type NestedStructBuilder struct {
    internal *NestedStruct
    errors map[string]cog.BuildErrors
}

func NewNestedStructBuilder() *NestedStructBuilder {
	resource := &NestedStruct{}
	builder := &NestedStructBuilder{
		internal: resource,
		errors: make(map[string]cog.BuildErrors),
	}

	builder.applyDefaults()

	return builder
}

func (builder *NestedStructBuilder) Build() (NestedStruct, error) {
	var errs cog.BuildErrors

	for _, err := range builder.errors {
		errs = append(errs, cog.MakeBuildErrors("NestedStruct", err)...)
	}

	if len(errs) != 0 {
		return NestedStruct{}, errs
	}

	return *builder.internal, nil
}

func (builder *NestedStructBuilder) StringVal(stringVal string) *NestedStructBuilder {
    builder.internal.StringVal = stringVal

    return builder
}

func (builder *NestedStructBuilder) IntVal(intVal int64) *NestedStructBuilder {
    builder.internal.IntVal = intVal

    return builder
}

func (builder *NestedStructBuilder) applyDefaults() {
}
